peer-ids/peer-ids.md

```diff
@@ -0,0 +1,117 @@
+# Spec:  Peer Ids and Keys
+
+## Keys
+
+
+Our key pairs are stored on disk using a simple protobuf defined in [libp2p/go-libp2p-crypto/pb/crypto.proto#L5](https://github.com/libp2p/go-libp2p-crypto/blob/master/pb/crypto.proto#L5):
+
+```protobuf
+enum KeyType {
+	RSA = 0;
+	Ed25519 = 1;
+	Secp256k1 = 2;
+	ECDSA = 3;
+}
+
+message PublicKey {
+	required KeyType Type = 1;
+	required bytes Data = 2;
+}
+
+message PrivateKey {
+	required KeyType Type = 1;
+	required bytes Data = 2;
+}
+```
+
+As should be apparent from the above code block, this proto simply encodes for transmission a public/private key pair along with an enum specifying the type of keypair.
+
+#### Where it's used?
+
+Keys are used in two places in libp2p.  The first is for signing messages.  Here are some examples of messages we sign:
+ - IPNS records
+ - PubSub messages (coming soon)
+ - SECIO handshake
+
+The second is for generating peer ids; this is discussed in the section below.
+
+## Peer Ids
+
+Here is the process by which we generate peer id's based on the public/private keypairs described above:
+
+  1. Encode the public key into the protobuf.
+  2. Serialize the protobuf containing the public key into bytes using the [canonical protobuf encoding](https://developers.google.com/protocol-buffers/docs/encoding).
+  3.  If the length of the serialized bytes <= 42, then we compute the "identity" multihash of the serialized bytes.  In other words, no hashing is performed, but the [multihash format is still followed](https://github.com/multiformats/multihash) (byte plus varint plus serialized bytes).  The idea here is that if the serialized byte array is short enough, we can fit it in a multihash proto without having to condense it using a hash function.
+  4. If the length is >42, then we hash it using it using the SHA256 multihash.
+
+For more information, refer to this block in [libp2p/go-libp2p-peer/peer.go](https://github.com/libp2p/go-libp2p-peer/blob/master/peer.go):
+
+```
+// MaxInlineKeyLength is the maximum length a key can be for it to be inlined in
+// the peer ID.
+//
+// * When `len(pubKey.Bytes()) <= MaxInlineKeyLength`, the peer ID is the
+//   identity multihash hash of the public key.
+// * When `len(pubKey.Bytes()) > MaxInlineKeyLength`, the peer ID is the
+//   sha2-256 multihash of the public key.
+const MaxInlineKeyLength = 42
+```
+
+
+## How Keys are Encoded and Messages Signed
+
+Four key types are supported:
+ - RSA
+ - Ed25519
+ - Secp256k1
+ - ECDSA
+
+Implementations SHOULD support RSA and Ed25519. Implementations MAY support Secp256k1 and ECDSA, but nodes using those keys may not be able to connect to all other nodes.
```

Comment:
I think SHOULD is correct here, because might be reasons not to support them, e.g. security or performance considerations.
If an implementer really wants to produce a useless implementation, he'll find plenty of ways to do so...

Reply:
That lack of definition is not good for a spec. If you leave this open ended, one implementor could choose to support RSA only, another Ed25519 only, and their implementations won't interoperate. Choosing _your_ private key type based on arbitrary criteria is a different topic than interoperability.

Reply:
Maybe a libp2p language implementation MUST support RSA and Ed25519, but we call out that an individual libp2p application or network can make its own call on what to support. So if you have security or performance concerns for your deployment, you can disable RSA and Ed25519 and use whatever you need.

If a language implementation doesn't happen to support both of the baseline keys, that's fine, but we would consider it a missing feature. Whereas we'd consider an implementation without Secp256k1 "feature complete", since that's not a MUST.

Either way, we should make it clear that not supporting them will limit your interoperability. Which might be fine for some use cases, as @marten-seemann says.

Reply:
I think overcomplicating this. These are specs for **implementors**; what the application using an implementation chooses or not to do is irrelevant here. 

My gripe is that, as an implementor, I look at this statement and it's totally useless because I don't know what to do (MUST), and where I'm given optionality (SHOULD/COULD), I don't know what the implications are.

Reply:
> what the application using an implementation chooses or not to do is irrelevant here.

The concrete issue is: we receive a handshake containing an RSA key. Is it mandatory to accept it? Both Rust and Substrate would be very glad if it wasn't.

Reply:
IMO Your implementation MUST support it. Your implementation MAY allow the user to enable/disable specific key types via configuration.

Reply:
But this spec is supposed to define what happens on the wire, because that's all that matters for compatibility.
To SHOULD is the right way to put it, then.

Reply:
@tomaka and I cycled ideas back and forth via chat, as it was easier. We concluded that:
1. Specs cover behaviour insofar interoperability is concerned.
2. Implementations MUST support a RSA and Ed25519, but just like transports are pluggable and one can configure ciphersuites in TLS, implementations MAY allow the user to configure which key types their host will support. So while RSA (source of concern) MUST be supported, the user can choose to disable it.
3. libp2p aims to put the choice in user’s hands, in the same way that e.g. Go ships with RSA but it’s up to the user to use it.

Thanks for everyone who participated.

Reply:
For an endpoint it’s indistinguishable if its peer supports RSA but disabled RSA support or if it uses an implementation that doesn’t support RSA in the first place. Therefore, I see little value in making it a requirement to support a certain encryption scheme.

Reply:
@marten-seemann this is about creating clarity for _implementors_ of the spec. Splintering a cloudy SHOULD into clear and detailed MUSTs and MAYs goes a long way here.

Reply:
It's not enforceable by the peer anyway, so whatever language we put into the documents, implementors will do what implementors see necessary to do. Requiring something that's not necessary for the function of the protocol seems like a wrong use of a normative MUST to me.

Reply:
If you read the original statement from the viewpoint of an implementer, you will have no clue what to do, as everything is optional. You'd end up guessing or digging into the code of other implementations, which defeats the purpose of writing a spec altogether.

Our specs are useless if they don't provide a baseline framework for interoperability. In this case, this is achieved by making all implementations support RSA and Ed25519.

Under this arrangement, if peer A (implementation X) shares an RSA key with peer B (implementation Y), and the latter rejects it, peer A can be certain that peer B has disabled support for RSA keys.